### PR TITLE
Feature/servers

### LINKED ISSUE #1050: Allow generation of servers and base paths in `api`
**Is your feature request related to a problem? Please describe.**
OpenAPI3 allows defining servers (https://swagger.io/docs/specification/api-host-and-base-path/) which can help define the scheme, description, and variables which form a server. This request is to enable spot to generate that for the OpenAPI Spec.

**Describe the solution you'd like**
Allow the definition of `server` inside the `@api` module of spot, with the following conversions:
```ts
@api({
  name: "My API",
  version: '1',
})
class Api {
  @servers
  server(
    @url
    url: 'https://{environment}.example.com/v2',
    @variables
    variables: {
      /** API description for server variables */
      environment: 'api' | 'api.dev' | 'api.staging'
    }
  ) {}
}
```

```yaml
servers:
  - url: https://{environment}.example.com/v2
    variables:
      environment:
        default: api 
        description: API description for environment variables
        enum:
          - api         
          - api.dev  
          - api.staging
```

**Describe alternatives you've considered**
Haven't thought of any other designs at this moment, likely will need to think about if it would be useful to add (a) a default value, which could be the `first` value?  and (b) multiple servers.

### PR DESCRIPTION
**Pull request by oben01**
## Description, Motivation and Context

Please see https://github.com/airtasker/spot/issues/1050
This feature allows to declare one or multiple server.

To declare a server : 
``` typescript
@api({ name: "contract" })
class Contract {
    /**
     * Production server
     */
    @oa3server({ url: "https://{username}.gigantic-server.com:{port}/{basePath}" })
    productionServer(
        /**
         * this value is assigned by the service provider, in this example `gigantic-server.com`
         */
        @oa3serverVariable
            username: String = "demo",
        @oa3serverVariable
            port: "8443" | "443" = "8443",
        @oa3serverVariable
            basePath: String = "v2"
    ) {}
}
```
To declare multiple servers : 
``` Typescript
@api({ name: "contract" })
class Contract {
    @oa3server({ url: "https://{username}.gigantic-server.com:{port}/{basePath}" })
    productionServer(
        /** this value is assigned by the service provider, in this example `gigantic-server.com` */
        @oa3serverVariable
            username: String = "demo",
        @oa3serverVariable
            port: "8443" | "443" = "8443",
        @oa3serverVariable
            basePath: String = "v2"
    ) {}
    @oa3server({ url: "https://{username}.gigantic-server.com:{port}/{basePath}" })
    devServer(
        /** this value is assigned by the service provider, in this example `gigantic-server.com` */
        @oa3serverVariable
            username: String = "dev",
        @oa3serverVariable
            port: "8080" | "80" = "8080",
        @oa3serverVariable
            basePath: String = "v2"
    ) {}
}
```

This is my first time dealing with Typescript. So Please feel free to give guidance to enhance the code.

I got some questions : 
- I used test to debug my code. Is that possible to add documentation about development process  (debug Locally, Generate a library locally, testing the cli ..... )(https://github.com/airtasker/spot/issues/1041) 

Thank you for your work 👍 

## Checklist:

- [x] I've added/updated tests to cover my changes
- [x] I've created an issue associated with this PR

### CODE REVIEW

**Comment by oben01** [comment]
The generated swagger look like : 
``` yaml
servers:
  - url: https://{username}.gigantic-server.com:{port}/{basePath}
    variables:
      username:
        default: '"demo"'
      port:
        default: '"8443"'
        enum:
          - '8443'
          - '443'
      basePath:
        default: '"v2"'
```

default string values are generated between brackets 

swagger

I need to dig to find a solution to parse  thoses values without brackets. I ll keep you posted.

**Comment by lfportal** [comment]
Hi @oben01 sorry the delay, I'll take a look at this over the next couple of days

**Comment by lfportal** [comment]
Hi @oben01, I've been a bit swamped, appreciate the work on this. I don't think it was quite clear on the original issue. The preferred approach that was settled on was https://github.com/airtasker/spot/issues/1050#issuecomment-705492666

```ts
@api({
  name: "My API",
  version: '1'
})
class Api {
  /**
   * The production API server
   */
  @oa3server({ url: "https://{username}.gigantic-server.com:{port}/{basePath}" })
  productionServer(
    @oa3serverVariables
    variables: {
      /**
       * this value is assigned by the service provider, in this example `gigantic-server.com`
       * 
       * @default "demo"
       */
      username: String,
      /**
       * @default "8443"
       */
      port: "8443" | "443",
      /**
       * @default "v2"
       */
      basePath: String
    }
  ) {}
}
```

Could you modify the PR for the above syntax? Sorry for not getting back to you earlier.

**Comment by oben01** [comment]
@lfportal I changed the PR to the preferred approach. Can you please take a look.

File: lib/src/generators/openapi3/openapi3.ts [hunk 1]
```diff
@@ -349,3 +356,53 @@ function exampleToOpenApiExampleSet(
...
+function contractToOa3ServerObject(
+  contract: Contract
+): ServerObject[] | undefined {
+  if (contract.oa3servers?.length === 0) {
+    return undefined;
+  }
+  const servers = contract.oa3servers;
+  const serversObject: ServerObject[] = [];
+  if (servers) {
+    for (const server of servers) {
```
Comment (lfportal) [inline lib/src/generators/openapi3/openapi3.ts]: If `servers` is `undefined`, we can just return `undefined`, then we can `.map` on `servers`

File: lib/src/parsers/oa3server-parser.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  const servers: Oa3Server[] = [];
+  for (const method of serverMethods) {
+    const serverResult = parseOa3Server(method, typeTable, lociTable);
+    if (serverResult.isErr()) return serverResult;
+    servers.push(serverResult.unwrap());
+  }
+  return ok(servers);
+}
+
+export function parseOa3Server(
```
Comment (lfportal) [inline lib/src/parsers/oa3server-parser.ts]: Is this function is anywhere else? If not, we can remove the `export` declaration

File: lib/src/parsers/oa3server-parser.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  return ok({
+    url: urlLiteral.getLiteralValue(),
+    description: description,
+    oa3ServerVariables: serverVariables
+  });
+}
+
+export function parseOa3Variables(
```
Comment (lfportal) [inline lib/src/parsers/oa3server-parser.ts]: Similar to above, we can remove the `export` declaration

File: lib/src/parsers/oa3server-parser.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  parameter: ParameterDeclaration,
+  typeTable: TypeTable,
+  lociTable: LociTable
+): Result<Oa3ServerVariable[], ParserError> {
+  // TODO: retrieve JsDoc as server variable description https://github.com/dsherret/ts-morph/issues/753
```
Comment (lfportal) [inline lib/src/parsers/oa3server-parser.ts]: Not sure I understand this `TODO` here. Was this for something else?

File: lib/src/parsers/oa3server-parser.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  getPropValueAsStringOrThrow
+} from "./parser-helpers";
+import { Oa3serverConfig } from "../syntax/oa3server";
+import { parseType } from "./type-parser";
+import { TypeTable } from "../types";
+import { LociTable } from "../locations";
+import { err, ok, Result } from "../util";
+import { OptionalNotAllowedError, ParserError } from "../errors";
+
+export function parseOa3Servers(
```
Comment (lfportal) [inline lib/src/parsers/oa3server-parser.ts:23]: I think we should add tests for this. There's some good examples in the `/lib/src/parsers/` directory.

**Review by lfportal — COMMENTED** [review]
@oben01 Overall approach looks good 👍 , just couple small comments.

File: lib/src/parsers/oa3server-parser.ts (hunk 4 above)
Comment (oben01) [reply lib/src/parsers/oa3server-parser.ts]: bad copy-paste

File: lib/src/parsers/oa3server-parser.ts (hunk 5 above)
Comment (oben01) [reply lib/src/parsers/oa3server-parser.ts:23]: Done

**Comment by oben01** [comment]
@lfportal Thank you for the review 🥇 
I just added oa3server-parser tests + some clean up according to your advice.
Please let me know if the code need more improvement. 👍

**Review by lfportal — APPROVED** [review]
Thanks for the quick turnaround!

**Comment by oben01** [comment]
@lfportal Could you please generate a new release ?

**Pull request merged**